tests/unit/s2n_alerts_protocol_test.c

```diff
@@ -0,0 +1,525 @@
+                EXPECT_SUCCESS(s2n_connection_set_blinding(client, S2N_SELF_SERVICE_BLINDING));
+                EXPECT_SUCCESS(s2n_connection_set_config(client, config));
+
+                DEFER_CLEANUP(struct s2n_test_io_stuffer_pair io_pair = { 0 }, s2n_io_stuffer_pair_free);
+                EXPECT_OK(s2n_io_stuffer_pair_init(&io_pair));
+                EXPECT_OK(s2n_connections_set_io_stuffer_pair(client, server, &io_pair));
+
+                if (type == S2N_TEST_DURING_HANDSHAKE) {
+                    /* Partially perform handshake */
+                    EXPECT_OK(s2n_negotiate_test_server_and_client_until_message(server, client, SERVER_CERT));
+                } else {
+                    /* Complete handshake */
+                    EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server, client));
+                }
+                EXPECT_EQUAL(server->actual_protocol_version, S2N_TLS13);
+
+                struct s2n_connection *receiver = server;
+                struct s2n_connection *sender = client;
+                if (mode == S2N_CLIENT) {
+                    receiver = client;
+                    sender = server;
+                }
+
+                /* Send alert */
+                struct s2n_blob alert = { 0 };
+                s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+                EXPECT_SUCCESS(s2n_blob_init(&alert, error_alerts[i], sizeof(error_alerts[0])));
+                EXPECT_OK(s2n_record_write(sender, TLS_ALERT, &alert));
+                EXPECT_SUCCESS(s2n_flush(sender, &blocked));
+
+                /**
+                 *= https://tools.ietf.org/rfc/rfc8446#section-6
+                 *= type=test
+                 *# Upon receiving an error alert, the TLS implementation
+                 *# SHOULD indicate an error to the application and MUST NOT allow any
+                 *# further data to be sent or received on the connection.
+                 */
+                if (type == S2N_TEST_DURING_HANDSHAKE) {
+                    EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(receiver, &blocked), S2N_ERR_ALERT);
+                    EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate(receiver, &blocked), S2N_ERR_CLOSED);
+                } else {
+                    EXPECT_FAILURE_WITH_ERRNO(s2n_recv(receiver, data, sizeof(data), &blocked), S2N_ERR_ALERT);
+                    EXPECT_EQUAL(s2n_recv(receiver, data, sizeof(data), &blocked), END_OF_DATA);
+                    EXPECT_FAILURE_WITH_ERRNO(s2n_send(receiver, data, sizeof(data), &blocked), S2N_ERR_CLOSED);
+                }
+
+                /**
+                 *= https://tools.ietf.org/rfc/rfc8446#section-6.2
+                 *= type=test
+                 *# Upon transmission or
+                 *# receipt of a fatal alert message, both parties MUST immediately close
+                 *# the connection.
+                 */
+                EXPECT_TRUE(s2n_connection_is_closed(receiver, S2N_FULL_CLOSED));
```

Comment:
Should the sender also be expected to be closed here?

Reply:
If the sender had actually encountered an error and sent an alert in response, yes. But this test is just testing receiving, so we fake the alerts in order to get a wider variety. The next test tests the sender side.